File: Tree/All_Nodes_Distance_K_in_Binary_Tree.py
#Problem link:- https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/

"""We are given a binary tree (with root node root), a target node, and an integer 
value K.Return a list of the values of all nodes that have a distance K from the 
target node.  The answer can be returned in any order."""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, K: int):
        import collections
        def dfs(root,par=None):
            if root:
                root.par = par
                dfs(root.left,root)
                dfs(root.right,root)
        dfs(root)
        queue = collections.deque([(target,0)])
        visited = {target}
        while queue:
            if queue[0][1] == K:
                return [node.val for node,d in queue]
            node,d = queue.popleft()
            for nei in (node.left,node.right,node.par):
                if nei and nei not in visited:
                    visited.add(nei)
                    queue.append((nei,d+1))
        return []

File: Tree/AllNodesDistanceKInBinaryTree.cpp
//Problem Link:- https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/

#include<bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void dfs(TreeNode* root,TreeNode* par,unordered_map<TreeNode*,TreeNode*>& parent){
        if (!root){
            return;
        }
        parent[root] = par;
        dfs(root->left,root,parent);
        dfs(root->right,root,parent);
    }
    
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        unordered_map<TreeNode*,TreeNode*> parent;
        dfs(root,NULL,parent);
        vector<int> result;
        queue<pair<TreeNode*,int>> pq;
        pq.push({target,0});
        unordered_set<TreeNode*> seen;
        seen.insert(target);
        while (!pq.empty()){
            auto x = pq.front();
            if (x.second == k){
                break;
            }
            pq.pop();
            TreeNode* node = x.first;
            int d = x.second;
            if (node->left && seen.count(node->left) == 0){
                pq.push({node->left,d+1});
                seen.insert(node->left);
            }
            if (node->right && seen.count(node->right) == 0){
                pq.push({node->right,d+1});
                seen.insert(node->right);
            }
            if (parent[node] != NULL && seen.count(parent[node]) == 0){
                pq.push({parent[node],d+1});
                seen.insert(parent[node]);
            }
        }
        while (!pq.empty()){
            pair<TreeNode*,int> pr = pq.front();
            pq.pop();
            result.push_back(pr.first->val);
        }
        return result;
    }
};